File: examples/governance_static_go/domain/ai_model.go
package domain

import (
	"time"
	"github.com/google/uuid"
	"fmt"
)

type AiModel struct {
	ID string `json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	Name string `json:"name"`
	Version string `json:"version"`
	ProviderId string `json:"provider_id"`
	Description string `json:"description"`
	RiskLevel string `json:"risk_level"`
	RegisteredAt time.Time `json:"registered_at"`
	ParentModelId string `json:"parent_model_id"`
	DerivationType string `json:"derivation_type"`
	Capabilities []Capability `json:"capabilities"`
	IntendedUses []IntendedUse `json:"intended_uses"`
	Status string `json:"status"`
}

func NewAiModel(name string, version string, providerId string, description string, riskLevel string, registeredAt time.Time, parentModelId string, derivationType string, capabilities []Capability, intendedUses []IntendedUse, status string) *AiModel {
	a := &AiModel{
		ID:        uuid.New().String(),
		Name: name,
		Version: version,
		ProviderId: providerId,
		Description: description,
		RiskLevel: riskLevel,
		RegisteredAt: registeredAt,
		ParentModelId: parentModelId,
		DerivationType: derivationType,
		Capabilities: capabilities,
		IntendedUses: intendedUses,
		Status: status,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
	return a
}

func (a *AiModel) Validate() error {
	if a.Name == "" {
		return &ValidationError{Field: "name", Message: "name can't be blank"}
	}
	if a.Version == "" {
		return &ValidationError{Field: "version", Message: "version can't be blank"}
	}
	if a.RiskLevel != "" {
		validRiskLevel := map[string]bool{"low": true, "medium": true, "high": true, "critical": true}
		if !validRiskLevel[a.RiskLevel] {
			return &ValidationError{Field: "risk_level", Message: fmt.Sprintf("risk_level must be one of: low, medium, high, critical, got: %s", a.RiskLevel)}
		}
	}
	if a.DerivationType != "" {
		validDerivationType := map[string]bool{"fine-tuned": true, "distilled": true, "retrained": true, "quantized": true}
		if !validDerivationType[a.DerivationType] {
			return &ValidationError{Field: "derivation_type", Message: fmt.Sprintf("derivation_type must be one of: fine-tuned, distilled, retrained, quantized, got: %s", a.DerivationType)}
		}
	}
	return nil
}

File: examples/governance/model_registry_domain/lib/model_registry_domain/ai_model/ai_model.rb
require 'hecks/mixins/model'

module ModelRegistryDomain
  class AiModel
    autoload :Capability, "model_registry_domain/ai_model/capability"
    autoload :IntendedUse, "model_registry_domain/ai_model/intended_use"
    autoload :Lifecycle, "model_registry_domain/ai_model/lifecycle"

    include Hecks::Model

    attribute :name
    attribute :version
    attribute :provider_id
    attribute :description
    attribute :risk_level
    attribute :registered_at
    attribute :parent_model_id
    attribute :derivation_type
    attribute :capabilities, default: [], freeze: true
    attribute :intended_uses, default: [], freeze: true
    attribute :status

    # State predicates — see lifecycle.rb for full state machine
    def draft?; status == "draft"; end
    def classified?; status == "classified"; end
    def approved?; status == "approved"; end
    def suspended?; status == "suspended"; end
    def retired?; status == "retired"; end

    VALID_RISK_LEVEL = ["low", "medium", "high", "critical"].freeze unless defined?(VALID_RISK_LEVEL)
    VALID_DERIVATION_TYPE = ["fine-tuned", "distilled", "retrained", "quantized"].freeze unless defined?(VALID_DERIVATION_TYPE)

    private

    def validate!
      raise ValidationError.new("name can't be blank", field: :name, rule: :presence) if name.nil? || (name.respond_to?(:empty?) && name.empty?)
      raise ValidationError.new("version can't be blank", field: :version, rule: :presence) if version.nil? || (version.respond_to?(:empty?) && version.empty?)
      if risk_level && !VALID_RISK_LEVEL.include?(risk_level)
        raise ValidationError, "risk_level must be one of: #{VALID_RISK_LEVEL.join(', ')}, got: #{risk_level}"
      end
      if derivation_type && !VALID_DERIVATION_TYPE.include?(derivation_type)
        raise ValidationError, "derivation_type must be one of: #{VALID_DERIVATION_TYPE.join(', ')}, got: #{derivation_type}"
      end
    end
  end
end
